geoopt/manifolds/poincare/__init__.py

```diff
@@ -57,60 +58,158 @@ def _check_point_on_manifold(self, x, *, atol=1e-5, rtol=1e-5):
     def _check_vector_on_tangent(self, x, u, *, atol=1e-5, rtol=1e-5):
         return True, None
 
-    def dist(self, x, y, *, keepdim=False):
-        return math.dist(x, y, c=self.c, keepdim=keepdim)
+    def dist(self, x, y, *, keepdim=False, dim=-1):
+        return math.dist(x, y, c=self.c, keepdim=keepdim, dim=dim)
 
-    def egrad2rgrad(self, x, u):
-        return math.egrad2rgrad(x, u, c=self.c)
+    def egrad2rgrad(self, x, u, *, dim=-1):
+        return math.egrad2rgrad(x, u, c=self.c, dim=dim)
 
-    def retr(self, x, u):
+    def retr(self, x, u, *, dim=-1):
         # always assume u is scaled properly
         approx = x + u
-        return math.project(approx, c=self.c)
+        return math.project(approx, c=self.c, dim=dim)
 
-    def projx(self, x):
-        return math.project(x, c=self.c)
+    def projx(self, x, dim=-1):
+        return math.project(x, c=self.c, dim=dim)
 
     def proju(self, x, u):
         return u
 
-    def inner(self, x, u, v=None, *, keepdim=False):
+    def inner(self, x, u, v=None, *, keepdim=False, dim=-1):
         if v is None:
             v = u
-        return math.inner(x, u, v, c=self.c, keepdim=keepdim)
+        return math.inner(x, u, v, c=self.c, keepdim=keepdim, dim=dim)
 
-    def expmap(self, x, u):
-        return math.project(math.expmap(x, u, c=self.c), c=self.c)
+    def norm(self, x, u, *, keepdim=False, dim=-1):
+        return math.norm(x, u, keepdim=keepdim, dim=dim)
 
-    def logmap(self, x, y):
-        return math.logmap(x, y, c=self.c)
+    def expmap(self, x, u, *, project=True, dim=-1):
+        res = math.expmap(x, u, c=self.c, dim=dim)
+        if project:
+            return math.project(res, c=self.c, dim=dim)
+        else:
+            return res
+
+    def logmap(self, x, y, *, dim=-1):
+        return math.logmap(x, y, c=self.c, dim=dim)
 
-    def transp(self, x, y, v, *more):
+    def transp(self, x, y, v, *more, dim=-1):
```

Comment:
I have concerns whether this method should work with non-matching `v, *more`. The easy way is to do this in a loop and do not care.

Reply:
Or maybe check and raise the warning?

Reply:
What is the point in `*more` anyway and where we will use it? I have completely missed that part.

Reply:
A flag in parameters may save nerves and add more clarity in what's going on

Reply:
like `stack=True`

Reply:
The point is to be able to transport multiple vectors with one pass, as this might be much more computationally cheaper. E.g. in Stiefel manifols this allows to performs this operation with one LU decomposition

Reply:
This is how flag is supposed to be implemented
```python
    def transp(self, x, y, v, *more, dim=-1, stack=True):
        if not more:
            return math.parallel_transport(x, y, v, c=self.c, dim=dim)
        else:
            if stack:
                vecs = torch.stack((v,) + more, dim=0)
                transp = math.parallel_transport(
                    x, y, vecs, c=self.c, dim=idx2sign(dim, x.dim())
                )
                return transp.unbind(0)
            else:
                return tuple(
                    math.parallel_transport(x, y, vec, c=self.c, dim=dim)
                    for vec in (v, *more)
                )

```

Reply:
However, tuple approach is not that bad 
```python
    def transp(self, x, y, v, *more, dim=-1):
        if not more:
            return math.parallel_transport(x, y, v, c=self.c, dim=dim)
        else:
            return tuple(
                math.parallel_transport(x, y, vec, c=self.c, dim=dim)
                for vec in (v, *more)
            )

```